pkg/cloud/vsphere/vsphere_cloud_config/config_test.go

```diff
@@ -0,0 +1,182 @@
+`
+
+func TestINIConfigConversion(t *testing.T) {
+
+	t.Run("basic config yaml conversion", func(t *testing.T) {
+		g := gmg.NewWithT(t)
+		configStruct, err := ReadConfig([]byte(basicConfigINI))
+		g.Expect(err).Should(gmg.Succeed())
+
+		convertedConfig, err := MarshalConfig(configStruct)
+		g.Expect(err).Should(gmg.Succeed())
+		// Trim left emptyline just for keep constant more readable
+		g.Expect(convertedConfig).To(gmg.BeEquivalentTo(strings.TrimLeft(basicConfigYaml, "\n")))
+	})
+
+	t.Run("basic config yaml conversion with vc section", func(t *testing.T) {
+		g := gmg.NewWithT(t)
+		configStruct, err := ReadConfig([]byte(basicConfigVcenterSectionINI))
+		g.Expect(err).Should(gmg.Succeed())
+
+		convertedConfig, err := MarshalConfig(configStruct)
+		g.Expect(err).Should(gmg.Succeed())
+		// Trim left emptyline just for keep constant more readable
+		g.Expect(convertedConfig).To(gmg.BeEquivalentTo(strings.TrimLeft(basicConfigVcenterSectionYAML, "\n")))
+	})
+
+	t.Run("Test multi DCs config yaml conversion", func(t *testing.T) {
+		g := gmg.NewWithT(t)
+		configStruct, err := ReadConfig([]byte(multiVCDCsConfigINI))
+		g.Expect(err).Should(gmg.Succeed())
+
+		g.Expect(len(configStruct.Vcenter)).To(gmg.Equal(3))
+
+		VC1 := configStruct.Vcenter["t1"]
+		g.Expect(len(VC1.Datacenters)).To(gmg.Equal(3))
+		g.Expect(VC1.Datacenters).To(gmg.BeComparableTo([]string{"DC0", "DC1", "DC2"}))
+		g.Expect(VC1.VCenterIP).To(gmg.Equal("10.0.0.1"))
+		g.Expect(VC1.SecretNamespace).To(gmg.Equal("kube-system"))
+		g.Expect(VC1.SecretName).To(gmg.Equal("tenant1-secret"))
+
+		VC2 := configStruct.Vcenter["10.0.0.2"]
+		g.Expect(len(VC2.Datacenters)).To(gmg.Equal(1))
+		g.Expect(VC2.Datacenters).To(gmg.BeComparableTo([]string{"DC3"}))
+		g.Expect(VC2.VCenterIP).To(gmg.Equal("10.0.0.2"))
+		g.Expect(VC2.IPFamilyPriority).To(gmg.Equal([]string{}))
+
+		VC3 := configStruct.Vcenter["10.0.0.3"]
+		g.Expect(len(VC3.Datacenters)).To(gmg.Equal(2))
+		g.Expect(VC3.Datacenters).To(gmg.BeComparableTo([]string{"DC5", "DC6"}))
+		g.Expect(VC3.IPFamilyPriority).To(gmg.Equal([]string{"ipv6"}))
+
+		_, err = MarshalConfig(configStruct)
+		g.Expect(err).Should(gmg.Succeed())
+	})
+
+	invalidConfigTestCases := []struct {
+		name         string
+		input        string
+		errSubstring string
+	}{
+		{
+			"rubbish", invalidConfig, "expected section header",
+		},
+		{
+			"bad port", invalidConfigWrongGlobalPort, "invalid global port parameter: parsed int bigger than zero",
+		},
+		{
+			"bad vc port", invalidConfigWrongVCPort, "invalid port parameter for vc 10.0.0.3: parsed int bigger than zero",
+		},
+		{
+			"empty", "", "vSphere config is empty",
+		},
+	}
+
+	for _, tc := range invalidConfigTestCases {
+		t.Run(fmt.Sprintf("test invalid config: %s", tc.name), func(t *testing.T) {
+			g := gmg.NewWithT(t)
+			_, err := ReadConfig([]byte(tc.input))
+			g.Expect(err).Should(gmg.HaveOccurred())
+			g.Expect(err.Error()).To(gmg.ContainSubstring(tc.errSubstring))
```

Comment:
You can shorten this, use `MatchError` to check the string value of the error, it will fail if no error occurs or if the error message doesn't match the next matcher
```suggestion
			g.Expect(err).To(gmg.MatchError(gmg.ContainSubstring(tc.errSubstring)))
```